### ISSUE TITLE
Add a method to add/insert empty columns

### ISSUE THREAD

**Issue by hhaensel** [JuliaData/DataFrames.jl#2783]
Currently adding empty columns must be done by hand AFAIK, e.g.
```
df.new = Vector{Union{Missing, Int}}(missing, nrow(df))
```
This is difficult to remember and certainly not easily accessible for beginners.
I propose to define
```
insertemptycols!(df::DataFrame, col::ColumnIndex, name_cols::Pair{<:Union{Symbol, AbstractString}, <:Union{DataType, Union}}...; makeunique::Bool=false) = 
    insertcols!(df, col, (name => Vector{Union{Missing, T}}(missing, nrow(df)) for (name, T) in name_cols)...; makeunique=makeunique)
    
insertemptycols!(df::DataFrame, name_cols::Pair{<:Union{Symbol, AbstractString}, <:Union{DataType, Union}}...; makeunique::Bool=false) = 
    insertemptycols!(df, ncol(df) + 1, name_cols...; makeunique=makeunique)
```
in order to provide, e.g.

```
julia> df = DataFrame(a=1:10);

julia> insertemptycols!(df, :new => Int)
10×2 DataFrame
 Row │ a      new    
     │ Int64  Int64?  
─────┼────────────────
   1 │     1  missing 
   2 │     2  missing 
   3 │     3  missing 
   4 │     4  missing 
   5 │     5  missing 
   6 │     6  missing 
   7 │     7  missing 
   8 │     8  missing 
   9 │     9  missing 
  10 │    10  missing 
```
I'll file a respective PR for testing. If this sounds like a reasonable feature, I'd be happy to also provide a docstring 😉

**Comment by bkamins** [comment]
```
insertcols!(hh, 2, [col => missings(Float64, nrow(hh)) for col in [:x, :y, :z]]...)
```
is indeed longer, but doable.

The key question is: do we want to add another function into the ecosystem? We try to minimize the number of verbs we provide to only these that are really needed and are not very niche. The issue is that we should answer ourselves if `insertemptycols!` would be:
1. used frequently enough (i.e. do people really create such columns in practice often? In what usage scenarios does it happen? as maybe the need comes from not knowing other ways to work with DataFrames.jl)
2. discovered by the users

to warrant its addition. Note that e.g. you were unaware of the `missings` function - this is exactly the problem we face.

I am leaving this question open, as I would some input from you and the community to judge this issue.

**Comment by hhaensel** [comment]
Fair enough!

I wouldn't have submitted the PR if I would have been aware of `missings()`. So I won't fight for it 😉

**Comment by bkamins** [comment]
Let us wait for others to comment.

CC @pdeffebach

**Comment by bkamins** [comment]
And thank you for the activity. We need such discussions to make sure we made a conscious decision.

**Comment by pdeffebach** [comment]
I was not aware of `missings`, but am happy it exists and agree that this probably doesn't merit a whole new function.

**Comment by hhaensel** [comment]
Just another thought; what about adding this functionality to `insertcols!()`?
The syntax could be
```
insertcols!(hh, :x => Float64)
```
But this would be breaking for anyone who wants to add a Datatype to a new column.
Alternatively, I could think of the following syntax
```
insertcols!(hh, :x => (Float64, missing), makeunique=true)     # insert column of type Float64? and assign `missing`
insertcols!(hh, :x => (Float64,), makeunique=true)             # insert column of type Float64? and assign `missing`
insertcols!(hh, :x => (Float64, 3), makeunique=true)           # insert column of type Float64 and assign the value 3
insertcols!(hh, :x => (Float64, 3, missing), makeunique=true)  # insert column of type Float64? and assign the value 3
```
The code would look like this
```
 if !(item isa AbstractVector)
            T = eltype(item)
            if item isa Tuple{T} where {T<:Union{DataType, Union}, v}
                T = item[1]
            elseif item isa Tuple{T, v} where {T<:Union{DataType, Union}, v}
                T, item = item
                item = missing
            elseif item isa Tuple{T, v, Missing} where {T<:Union{DataType, Union}, v}
                T = Union{Missing, item[1]}
                item = item[2]
            end
            if item isa Missing
                item_new = missings(T, target_row_count)
            elseif item isa Union{AbstractArray{<:Any, 0}, Ref}
                item_new = fill!(Tables.allocatecolumn(T, target_row_count), x)
            else
                @assert !(item isa AbstractArray)
                item_new = fill!(Tables.allocatecolumn(T, target_row_count), item)
            end
```

**Comment by kingasiuta** [comment]
I think it is not needed you can just do:
```
julia> df = DataFrame(a=[1,2,3])
3×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
  ⋮  │   ⋮
3 rows omitted

julia> insertcols!(df, :b => missings(Int))
3×2 DataFrame
 Row │ a      b       
     │ Int64  Missing 
─────┼────────────────
   1 │     1  missing 
   2 │     2  missing 
   3 │     3  missing 

julia> df[!, :c] .= missings(Float64)
3-element Array{Missing,1}:
 missing
 missing
 missing

julia> df
3×3 DataFrame
 Row │ a      b        c       
     │ Int64  Missing  Missing 
─────┼─────────────────────────
   1 │     1  missing  missing 
   2 │     2  missing  missing 
   3 │     3  missing  missing 
```

**Comment by bkamins** [comment]
In general it would be better to use:
```
julia> missings(Int)
0-dimensional Array{Union{Missing, Int64}, 0}:
missing
```

and all would work nicely except that currently it does not because in Julia Base broadcasting does type narrowing always, e.g.:
```
julia> (1:10) .=> Union{Int,Missing}[1:10;]
10-element Vector{Pair{Int64, Int64}}:
  1 => 1
  2 => 2
  3 => 3
  4 => 4
  5 => 5
  6 => 6
  7 => 7
  8 => 8
  9 => 9
 10 => 10
```
which is a bit unfortunate in this case (though it has its benefits).

I am not sure what would be the best thing to do. For sure what we could do is in pseudo broadcasting of `Ref` or `0`-dimensional arrays we could avoid making type narrowing (and document it since already pseudo broadcasting is not 100% consistent with broadcasting anyway). But I am not sure we want to go this way. @nalimilan - any opinion here?

**Comment by hhaensel** [comment]
The fact that `missings(Int)` creates a 0-dimensional array is really nice. (Something more, I learned today!)
Looking at the respective code I learned how to create a 0-dimensional array with a Union-type 😄 , something that I was trying to find out in the past.

Interestingly, part of my code provides a workaround: Determining `T = eltype(item)` beforehand produces the desired result:
```
if !(item isa AbstractVector)
            T = eltype(item)
            if item isa Union{AbstractArray{<:Any, 0}, Ref}
                @info "hello"
                item_new = fill!(Tables.allocatecolumn(T, target_row_count), item[])
            else
                @assert !(item isa AbstractArray)
                item_new = fill!(Tables.allocatecolumn(T, target_row_count), item)
            end
        elseif item isa AbstractRange
```
e.g.
```
julia> hh = DataFrame(a=1:10, b = 10:10:100);

julia> insertcols!(hh, :c => missings(Float64) , makeunique=true)
[ Info: hello
10×3 DataFrame
 Row │ a      b      c
     │ Int64  Int64  Float64?
─────┼────────────────────────
   1 │     1     10   missing
   2 │     2     20   missing
   3 │     3     30   missing
   4 │     4     40   missing
   5 │     5     50   missing
   6 │     6     60   missing
   7 │     7     70   missing
   8 │     8     80   missing
   9 │     9     90   missing
  10 │    10    100   missing
```

**Comment by hhaensel** [comment]
This led me to thinking about a possible enhancement of `fill()`
```
julia> import Base.fill

julia> fill(v, ::Type{T}, dims...) where T = fill!(Array{T}(undef, dims), v)
fill (generic function with 5 methods)

julia> fill(5, Number)
0-dimensional Array{Number, 0}:
5

julia> fill(5, Number, 3)
3-element Vector{Number}:
 5
 5
 5
```

**Comment by bkamins** [comment]
>  Determining `T = eltype(item)` beforehand produces the desired result:

Yes - this is what I am talking about. The point is that doing this would break the broadcasting contract in Julia Base, so we need to make a conscious decision to go this way (actually if we did this I would do it in other place in the code to ensure that this change is propagated consistently in the whole DataFrames.jl package).

> a possible enhancement of `fill`

Again - agreed. I was also thinking about it, but it would have to be a change in Julia Base and would not solve our problem unfortunately.

---

In summary: I think we are pretty close to the solution. The main decision for us to make is if we want to allow "pseudo broadcasting" to diverge from Julia Base broadcasting in this area of application.

**Comment by hhaensel** [comment]
Ah, thanks for explaining, I didn't get your comment right, then.

**Comment by bkamins** [comment]
Just for your reference, even without `missings` one could just write (assuming we allow the "pseudo broadcasting" design change) `Ref{Union{Int,Missing}}(missing)` using the Julia Base syntax. Then you can write ` Ref{Union{Int,Missing}}(whatever_value)` flexibly.

**Comment by bkamins** [comment]
>  I didn't get your comment right, then.

Sorry for being brief. What you proposed in your code is exactly changing the way how pseudo broadcasting works to adjust it to our case. In my opinion `eltype` narrowing in this case is unwanted, but on the other hand:
```
identity.(collection)
```
is a standard idiom for narrowing down the `eltype` of a collection in Julia Base.

The problem is that we cannot distinguish between the two, as internally e.g. `df[!, :b] .= v` will inject `identity` transformation by default to the code. Therefore we have (this is standard broadcasting):
```
julia> df = DataFrame(a=1:3)
3×1 DataFrame
 Row │ a
     │ Int64
─────┼───────
   1 │     1
   2 │     2
   3 │     3

julia> df[!, :b] .= Any["a"]
3-element Vector{String}:
 "a"
 "a"
 "a"

julia> df
3×2 DataFrame
 Row │ a      b
     │ Int64  String
─────┼───────────────
   1 │     1  a
   2 │     2  a
   3 │     3  a
```

**Comment by hhaensel** [comment]
Thank you for the detailed comments. I understand most of what you wrote, however, I can't get your example working
```
insertcols!(hh, :c => Ref{Union{Missing, Float64}}(missing) , makeunique=true)
```
results in
```
10×3 DataFrame
 Row │ a      b      c
     │ Int64  Int64  Missing
─────┼───────────────────────
   1 │     1     10  missing
   2 │     2     20  missing
   3 │     3     30  missing
   4 │     4     40  missing
   5 │     5     50  missing
   6 │     6     60  missing
   7 │     7     70  missing
   8 │     8     80  missing
   9 │     9     90  missing
  10 │    10    100  missing
```
Maybe, I still got something wrong.

On the other hand, I think there is good reason to use the `eltype(item)` rather than `typeof(item[])`.
1. If you pass a vector, then no narrowing is applied:
```
julia> hh = DataFrame(a=1:10, b = 10:10:100);

julia> a = Array{Number}(collect(1:10));

julia> insertcols!(hh, :c => a , makeunique=true)
10×3 DataFrame
 Row │ a      b      c
     │ Int64  Int64  Number
─────┼──────────────────────
   1 │     1     10       1
   2 │     2     20       2
   3 │     3     30       3
   4 │     4     40       4
   5 │     5     50       5
   6 │     6     60       6
   7 │     7     70       7
   8 │     8     80       8
   9 │     9     90       9
  10 │    10    100      10
```
2. If we'd opt for the change any user had the choice to either use the attached or the narrowed type
```
hh = DataFrame(a=1:10, b = 10:10:100);
a = fill!(Array{Number}(undef),1)

# insert the attached type (Number)
insertcols!(hh, :c1 => a);

# insert the narrowed type (Int)
insertcols!(hh, :c2 => a[])
```
results in
```
10×4 DataFrame
  Row │ a      b      c1      c2    
      │ Int64  Int64  Number  Int64 
─────┼─────────────────────────────
   1 │     1     10       1      1
   2 │     2     20       1      1
   3 │     3     30       1      1
   4 │     4     40       1      1
   5 │     5     50       1      1
   6 │     6     60       1      1
   7 │     7     70       1      1
   8 │     8     80       1      1
   9 │     9     90       1      1
  10 │    10    100       1      1
```
3. Narrowing for `missing` leads to type `Any` instead of `Missing`
4. I don't see any scenario, where this change would lead to unexpected results. But maybe, there is one?

**Comment by bkamins** [comment]
> I can't get your example working

It is not working now - it would be working if we adopted the change we are discussing.

> If you pass a vector, then no narrowing is applied:

Yes, that is why it is called "pseudo broadcasting".
The other inconsistency we have is that 1-element vectors are not expanded (and broadcasting expands them).

> `# insert the attached type (Number)`
`insertcols!(hh, :c1 => a);`
>
> `# insert the narrowed type (Int)`
`insertcols!(hh, :c2 => a[])`

Yes, I am aware of this and that is why I say we could consider to change the current behavior since pseudo broadcasting already deviates from broadcasting.

**Comment by hhaensel** [comment]
I'm getting to it. Thanks for your patient explanations. I am learning a lot 😉.

**Comment by bkamins** [comment]
I have thought about this issue. Allowing `missings(Int)` would be doable (we would need to change broadcasting assignment logic, `DataFrame` constructor, `insertcols!`, `combine`, `select`, `select!`, `transform` and `transform!` both for `AbstractDataFrame` and `GroupedDataFrame` where applicable), but as @nalimilan commented it breaks a standard pattern from Julia Base that type narrowing is performed when collections are created.

Therefore I think for now something like `df.x = missings(Int, nrow(df))` should be used, and `df.x .= missings(Int)` will do type narrowing as it does now.

In conclusion I close this issue. But if you strongly disagree please comment and we can re-open and discuss.

**Comment by bkamins** [comment]
Just as a reference in DataFrames.jl 1.3 the following will also work (not to be used of course, but just as an exercise):
```
julia> df = DataFrame(a=1:3)
3×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
   1 │     1 
   2 │     2 
   3 │     3 

julia> leftjoin!(df, DataFrame(a=[], x=Int[]), on=:a)
3×2 DataFrame
 Row │ a      x       
     │ Int64  Int64?  
─────┼────────────────
   1 │     1  missing 
   2 │     2  missing 
   3 │     3  missing 

julia> df = DataFrame(a=1:3)
3×1 DataFrame
 Row │ a     
     │ Int64 
─────┼───────
   1 │     1
   2 │     2
   3 │     3

julia> insertcols!(view(df, 1:0, :), :x => 1)
0×2 SubDataFrame

julia> df
3×2 DataFrame
 Row │ a      x       
     │ Int64  Int64?  
─────┼────────────────
   1 │     1  missing 
   2 │     2  missing 
   3 │     3  missing 
```

**Comment by hhaensel** [comment]
I'm fine with the decision.
I learned a lot through this discussion and I will come back here everytime I forgot how to insert columns, now that I have so many possibilities 😉
